vnc/vnc.c

```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+             * only got one screen, we will preserve their screen ID
+             * and any flags.  This may prevent us sending an unwanted
+             * SetDesktopSize message if the screen dimensions are
+             * a match. We can't do this with more than one screen,
+             * as we have no way to map different IDs
+             */
+            if (layout.count == 1 && v->client_layout.count == 1)
+            {
+                log_message(LOG_LEVEL_DEBUG, "VNC "
+                            "setting screen id to %d from server",
+                            layout.s[0].id);
+
+                v->client_layout.s[0].id = layout.s[0].id;
+                v->client_layout.s[0].flags = layout.s[0].flags;
+            }
+
+            if (vnc_screen_layouts_equal(&layout, &v->client_layout))
+            {
+                log_message(LOG_LEVEL_DEBUG, "Server layout is the same "
+                            "as the client layout");
+                v->initial_resize_status = VRS_DONE;
+            }
+            else
+            {
+                log_message(LOG_LEVEL_DEBUG, "Sending SetDesktopSize "
+                            "to change server layout");
+                error = send_set_desktop_size(v, &v->client_layout);
+                v->initial_resize_status = VRS_WAITING_FOR_RESIZE_CONFIRM;
+            }
+        }
+        else
+        {
+            log_message(LOG_LEVEL_DEBUG, "VNC server does not support resizing");
+
+            /* Force client to same size as server */
+            log_message(LOG_LEVEL_DEBUG, "Resizing client to server %dx%d",
+                        v->server_width, v->server_height);
+            error = resize_client(v, 0, v->server_width, v->server_height);
+            v->initial_resize_status = VRS_DONE;
+        }
+
+        g_free(layout.s);
+    }
+
+    if (error == 0)
+    {
+        error = send_update_request_for_resize_status(v);
+    }
+
+    return error;
+}
+
+/**************************************************************************//**
+ * Looks for a resize confirm in a framebuffer update request
+ *
+ * If the server supports resizes from us, this is used to find the
+ * reply to our initial resize request. See The RFB community wiki for details.
+ *
+ * @param v VNC object
+ * @return != 0 for error
+ */
+static int
+lib_framebuffer_waiting_for_resize_confirm(struct vnc *v)
+{
+    int error;
+    struct vnc_screen_layout layout = {0};
+    int response_code;
+
+    error = find_matching_extended_rect(v,
+                                        rect_is_reply_to_us,
+                                        NULL,
+                                        &response_code,
+                                        &layout);
+    if (error == 0)
+    {
+        if (layout.s != NULL)
+        {
+            if (response_code == 0)
+            {
+                log_message(LOG_LEVEL_DEBUG, "VNC server successfully resized");
```

Comment:
I think here's the place to log NewLayout at INFO level because resizing has been finished here, right?

Reply:
Agreed